profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSnapshotResultReusableBuffer.h

```diff
@@ -147,77 +139,42 @@ inline std::uint64_t StackSnapshotResultBuffer::SetSpanId(std::uint64_t value)
     return prevValue;
 }
 
-inline std::uint16_t StackSnapshotResultBuffer::GetFramesCount(void) const
-{
-    return _currentFramesCount;
-}
-
-inline StackSnapshotResultFrameInfo& StackSnapshotResultBuffer::GetFrameAtIndex(std::uint16_t index) const
+inline std::size_t StackSnapshotResultBuffer::GetFramesCount(void) const
 {
-    if (index >= _currentFramesCount || _stackFrames == nullptr)
-    {
-        StackSnapshotResultReusableBuffer::UnusableFrameInfo = StackSnapshotResultFrameInfo(static_cast<StackFrameCodeKind>(0xFFFF),
-                                                                                            static_cast<FunctionID>(0xFFFFFFFFFFFFFFFF),
-                                                                                            static_cast<UINT_PTR>(0xFFFFFFFFFFFFFFFF),
-                                                                                            static_cast<std::uint64_t>(0xFFFFFFFFFFFFFFFF));
-        return UnusableFrameInfo;
-    }
-
-    return *(_stackFrames + index);
+    return _instructionPointers.size();
 }
 
 inline void StackSnapshotResultBuffer::CopyInstructionPointers(std::vector<std::uintptr_t>& ips) const
 {
-    ips.reserve(_currentFramesCount);
-    for (size_t i = 0; i < _currentFramesCount; i++)
-    {
-        ips.push_back(_stackFrames[i].GetNativeIP());
-    }
+    ips.reserve(_instructionPointers.size());
+
+    // copy the instruction pointer to the out-parameter
+    ips = _instructionPointers;
 }
 
 // ----------- ----------- ----------- ----------- ----------- ----------- ----------- ----------- -----------
 
-inline std::uint16_t StackSnapshotResultReusableBuffer::GetCapacity()
+inline bool StackSnapshotResultReusableBuffer::AddFrame(std::uintptr_t ip)
 {
-    return _currentCapacity;
-}
+    const auto nextIdx = _instructionPointers.size();
 
-inline std::uint16_t StackSnapshotResultReusableBuffer::GrowCapacityAtNextReset()
-{
-    _nextResetCapacity = _currentCapacity + StackSnapshotResultReusableBuffer::MaxSnapshotStackDepth_IncStep;
-    if (_nextResetCapacity > StackSnapshotResultReusableBuffer::MaxSnapshotStackDepth_Limit)
-    {
-        _nextResetCapacity = StackSnapshotResultReusableBuffer::MaxSnapshotStackDepth_Limit;
-    }
-
-    return _nextResetCapacity;
-}
-
-inline bool StackSnapshotResultReusableBuffer::TryAddNextFrame(StackSnapshotResultFrameInfo** frameInfo, bool* hasCapacityForSubsequentFrames)
-{
-    if (_currentFramesCount >= _currentCapacity)
+    if (nextIdx >= MaxSnapshotStackDepth_Limit)
     {
         return false;
     }
 
-    if (frameInfo != nullptr)
-    {
-
-        *frameInfo = (_stackFrames + _currentFramesCount);
-    }
-
-    _currentFramesCount++;
-
-    if (hasCapacityForSubsequentFrames != nullptr)
+    const auto lastIdx = MaxSnapshotStackDepth_Limit - 1;
+    if (nextIdx == lastIdx)
     {
-
-        *hasCapacityForSubsequentFrames = (_currentFramesCount < _currentCapacity);
+        _instructionPointers.push_back(0);
```

Comment:
As discussed, it might be good to set MaxSnapshotStackDepth_Limit to +1 of the limit (i.e. 513 instead of 512) to avoid the "impossible" case of an exact 512 frames callstack

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/2750/commits/57c8be29c474387816ad3300189248aab62a7818